packages/controls/src/widget_dom_events.ts

```diff
@@ -0,0 +1,276 @@
+                    break
+                case "mouse":
+                    let prevent_default = this.get('prevent_default_action')
+                    let handler = this._dom_click.bind(this, view)
+                    view.el.addEventListener(event, handler)
+                    // Keep track of the listeners we are attaching so that we can
+                    // remove them if needed.
+                    this._cache_listeners(event, view, handler)
+                    break
+                default:
+                    console.log('Not familiar with that message source')
+                    break
+            }
+        }
+    }
+
+    attach_listeners() {
+        let current_source = this.get('source')
+        _.each(current_source.views, (view_promise) => {
+            Promise.resolve(view_promise).then((view) => {
+                this._add_listeners_to_view(view)
+            })
+        })
+    }
+
+    _add_key_listener(event_type, view) {
+        // Key listeners should:
+        //     + Only fire when the mouse is over the element.
+        //     + Not propagate up to the notebook (because imagine you
+        //       press 'x' on your widget and cut the cell...not what
+        //       you probably want the user to experience)
+        //
+        // The approach here is to add a key listener on mouseenter and remove
+        // it on mouseleave.
+        let key_handler = (event) => {
+            // console.log('Key presses FTW!', event)
+            if (this.get('ignore_modifier_key_events') &&
+               _.contains(this.get('_modifier_keys'), event.key)) {
+                // If the key event is supposed to be ignored, then skip it.
+                return
+            }
+            this._send_dom_event(event)
+            // Need this (and useCapture in the listener) to prevent the keypress
+            // from propagating to the notebook.
+            event.stopPropagation()
+            event.preventDefault()
+        }
+        // Last argument useCapture needs to be true to prevent the event from
+        // passing through to the notebook; also need to stopPropagation in key_handler.
+        let capture_event = true
+        let enable_key_listen = () => {document.addEventListener(event_type, key_handler, capture_event)}
+        let disable_key_listen = () => {document.removeEventListener(event_type, key_handler, capture_event)}
+        view.el.addEventListener('mouseenter', enable_key_listen)
+        view.el.addEventListener('mouseleave', disable_key_listen)
+        this._cache_listeners('mouseenter', view, enable_key_listen)
+        this._cache_listeners('mouseleave', view, disable_key_listen)
+    }
+
+
+    _dom_click(generating_view, event) {
+        // Get coordinates relative to the
+        let relative_xy = _get_position(generating_view, event)
+        event['relativeX'] = relative_xy.x
+        event['relativeY'] = relative_xy.y
+        if ('_array_xy' in generating_view) {
```

Comment:
I find the names _array_xy/arrayX a bit unclear. A natural name would maybe be natural_x and natural_y (no pun indented 😃 ) . Also would be useful to have more coordinates, I would say x, y are wrt the element, page_x,page_y would be relative to the page (including scrolling), window_x/y would be wrt upper left corner (excluding scolling) (I'm not 100% sure that is the same semantics as in js/css/html).
Also, I'd think normalize versions (between 0 and 1) would be useful.

Reply:
> Also would be useful to have more coordinates, I would say x, y are wrt the element, page_x,page_y would be relative to the page (including scrolling), window_x/y would be wrt upper left corner (excluding scolling) 

I think almost all currently returned in the event (full list is https://github.com/jupyter-widgets/ipywidgets/pull/1727/files#diff-6856e80b321629aaabf83b82a9939b43R23)

>  A natural name would maybe be natural_x and natural_y (no pun indented 😃 ) .

I'm fine with a different name...I also thought about `_image_x/imageX` but `natural` works (and puns are always in-scope).

> Also, I'd think normalize versions (between 0 and 1) would be useful.

Normalized versions of the natural coordinates, I assume?